source/opt/loop_utils.cpp

```diff
@@ -0,0 +1,379 @@
+  }
+
+ private:
+  // Return the basic block that |instr| belongs to.
+  ir::BasicBlock* GetParent(ir::Instruction* instr) {
+    return context_->get_instr_block(instr);
+  }
+
+  // Return the new def to use for the basic block |bb_id|.
+  // If |bb_id| does not have a suitable def to use then we:
+  //   - return the common def used by all predecessors;
+  //   - if there is no common def, then we build a new phi instr at the
+  //     beginning of |bb_id| and return this new instruction.
+  const ir::Instruction& GetOrBuildIncoming(uint32_t bb_id) {
+    assert(cfg_->block(bb_id) != nullptr && "Unknown basic block");
+
+    ir::Instruction*& incoming_phi = bb_to_phi[bb_id];
+    if (incoming_phi) {
+      return *incoming_phi;
+    }
+
+    // Check if one of the loop exit basic block dominates |bb_id|.
+    for (const ir::BasicBlock* e_bb : exit_bb_) {
+      if (dom_tree_.Dominates(e_bb->id(), bb_id)) {
+        incoming_phi = bb_to_phi[e_bb->id()];
+        assert(incoming_phi && "No closing phi node ?");
+        return *incoming_phi;
+      }
+    }
+
+    // Process parents, they will returns their suitable phi.
+    // If they are all the same, this means this basic block is dominated by a
+    // common block, so we won't need to build a phi instruction.
+    std::vector<uint32_t> incomings;
+    for (uint32_t pred_id : cfg_->preds(bb_id)) {
+      incomings.push_back(GetOrBuildIncoming(pred_id).result_id());
+      incomings.push_back(pred_id);
+    }
+    uint32_t first_id = incomings.front();
+    size_t idx = 0;
+    for (; idx < incomings.size(); idx += 2)
+      if (first_id != incomings[idx]) break;
+
+    if (idx >= incomings.size()) {
+      incoming_phi = bb_to_phi[incomings[1]];
+      return *incoming_phi;
+    }
+
+    // We have at least 2 definitions to merge, so we need a phi instruction.
+    ir::BasicBlock* block = cfg_->block(bb_id);
+
+    opt::InstructionBuilder<> builder(context_, &*block->begin());
+    incoming_phi = builder.AddPhi(insn_type_, incomings);
+
+    rewritten.insert(incoming_phi);
+
+    return *incoming_phi;
+  }
+
+  ir::IRContext* context_;
+  ir::CFG* cfg_;
+  const opt::DominatorTree& dom_tree_;
+  uint32_t insn_type_;
+  std::unordered_map<uint32_t, ir::Instruction*> bb_to_phi;
+  std::unordered_set<ir::Instruction*> rewritten;
+  const std::unordered_set<ir::BasicBlock*>& exit_bb_;
+};
+}  // namespace
+
+void LoopUtils::CreateLoopDedicateExits() {
+  ir::Function* function = loop_->GetHeaderBlock()->GetParent();
+  ir::LoopDescriptor& loop_desc = *context_->GetLoopDescriptor(function);
+  ir::CFG& cfg = *context_->cfg();
+  opt::analysis::DefUseManager* def_use_mgr = context_->get_def_use_mgr();
+
+  constexpr ir::IRContext::Analysis PreservedAnalyses =
+      ir::IRContext::kAnalysisDefUse |
+      ir::IRContext::kAnalysisInstrToBlockMapping;
+
+  // Gathers the set of basic block that are not in this loop and have at least
+  // one predecessor in the loop and one not in the loop.
```

Comment:
"and one not in the loop" does sound correct for the definition of an exit block.

Reply:
It is but the exit is not dominated by the loop so that can only happen in non-structured context:
```
if (cond) {
 for () {}
}
// merge-bb
```
It would remain valid to have the loop exit to be `merge-bb`, so it would have 2 incoming edges: from the if and from the loop.